# Working with titanic dataset

import all library that are going to be used

In [1]:
import pandas as pd
import numpy as np

import titanic dataset with tab saparator, and try to display sample data

In [2]:
data = pd.read_csv("titanic.tsv", sep='\t')
data.head()

PassengerId  Survived  Pclass  \
0            1         0     3.0   
1            2         1     1.0   
2            3         1     3.0   
3            4         1     1.0   
4            5         0     3.0   

                                                Name     Sex Age  SibSp Parch  \
0                            Braund, Mr. Owen Harris    male  22      1     0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38      1     0   
2                             Heikkinen, Miss. Laina  female  26      0     0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35      1     0   
4                           Allen, Mr. William Henry    male  35      0     0   

             Ticket     Fare Cabin Embarked     ship  
0         A/5 21171     7,25   NaN        S  Titanic  
1          PC 17599  71,2833   C85        C  Titanic  
2  STON/O2. 3101282    7,925   NaN        S  Titanic  
3            113803     53,1  C123        S  Titanic  
4            373450     8,05   NaN        S  Titanic

check datatypes to check if the data is in correct types

In [3]:
data.dtypes

PassengerId      int64
Survived         int64
Pclass         float64
Name            object
Sex             object
Age             object
SibSp            int64
Parch           object
Ticket          object
Fare            object
Cabin           object
Embarked        object
ship            object
dtype: object

from dtypes function we know that Pclass, Age, Parch and Fare column are in object type, so we should change it to numeric datatype

In [4]:
data['Pclass'] = pd.to_numeric(data.Pclass, errors='coerce')
data['Age'] = pd.to_numeric(data.Age, errors='coerce')
data['Parch'] = pd.to_numeric(data.Parch, errors='coerce')

data['Fare'] = data['Fare'].str.replace(',','.')
data['Fare'] = pd.to_numeric(data.Fare, errors='coerce')

data.dtypes

PassengerId      int64
Survived         int64
Pclass         float64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch          float64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
ship            object
dtype: object

We will try to describe data to understand it

In [5]:
data.describe()

PassengerId    Survived      Pclass          Age       SibSp  \
count   894.000000  894.000000  894.000000   696.000000  894.000000   
mean    446.030201    0.381432    2.305369    36.203448    0.604027   
std     259.208003    0.508529    0.847653   167.817961    2.571231   
min     -12.000000   -4.000000   -2.000000   -12.000000    0.000000   
25%     223.250000    0.000000    2.000000    20.000000    0.000000   
50%     444.500000    0.000000    3.000000    28.000000    0.000000   
75%     668.750000    1.000000    3.000000    38.000000    1.000000   
max    1143.000000    1.000000    3.000000  4435.000000   70.000000   

            Parch        Fare  
count  893.000000  892.000000  
mean     0.371781   32.075985  
std      0.768325   49.868844  
min      0.000000  -90.000000  
25%      0.000000    7.895800  
50%      0.000000   14.454200  
75%      0.000000   30.771850  
max      5.000000  512.329200

Locate Missing Data using isnull function

In [6]:
data.isnull()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   
891        False     False   False  False  False  False  False  False   False   
892        False     False   False  False  False  False  False  False   False   
893        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked   ship  
0    False   True     False  False  
1    False  False     False  False  
2    False   True     False  False  
3    False  False     False  False  
4    False   True     False  False  
..     ...    ...       ...    ...  
889  False  False     False  False  
890  False   True     False  False  
891  False  False     False  False  
892  False  False     False  False  
893  False   True     False  False  

[894 rows x 13 columns]

expand the missing data of each feature, from below result we know there are 6 attributes with missing value (Age, Parch, Ticket, Fare, Cabin, Embarked)

In [7]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            198
SibSp            0
Parch            1
Ticket           1
Fare             2
Cabin          686
Embarked         2
ship             0
dtype: int64

## Missing values

There are two basic options regarding missing data. 1) Drop the data or, 2) Input missing data.

### Cabin

Because we have 894 rows and there are 686 missing values in Cabin, so we can drop it.

In [8]:
data.drop("Cabin", inplace =True, axis =1)

In [9]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0     3.0   
1            2         1     1.0   
2            3         1     3.0   
3            4         1     1.0   
4            5         0     3.0   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked     ship  
0    0.0         A/5 21171   7.2500        S  Titanic  
1    0.0          PC 17599  71.2833        C  Titanic  
2    0.0  STON/O2. 3101282   7.9250        S  Titanic  
3    0.0            113803  53.1000        S  Titanic  
4    0.0            373450   8.0500        S  Titanic

### Age

We can input missing age data with the median value of the age

In [10]:
data['Age'].fillna(np.nanmedian(data['Age']), inplace = True)

### Parch

We can input missing parch data with the mean value of the parch

In [27]:
data['Parch'] = data['Parch'].fillna(np.mean(data['Parch']))

### Ticket

We can input missing ticket data with the mode value of the ticket

In [28]:
data['Ticket'].fillna(data['Ticket'].mode()[0], inplace=True)

### Fare

We can input missing fare data with the mean value of the fare

In [29]:
data['Fare'] = data['Fare'].fillna(np.mean(data['Fare']))

### Embarked

We can input missing embarked data with the mode value of the embarked

In [30]:
data['Embarked'].fillna(data['Embarked'].mode()[0], inplace=True)

Check missing value

In [32]:
data.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
ship           0
dtype: int64

### Saving new data

In [ ]:
data.to_csv('titanic_clean.csv')